server/services/discovery/discovery_service.py

```
@@ -283,24 +303,43 @@ def run_discovery_for_user(user_id, connected_providers):
+                # Serverless errors are already provider-prefixed (e.g. "aws/fn-name")
+                # so we can attribute them accurately.
+                for err in serverless_result["errors"]:
+                    for pname in ("gcp", "aws", "azure"):
+                        if pname in connected_providers and pname in err.lower():
+                            summary["provider_errors"].setdefault(pname, []).append(err)
+                            break
             logger.info(f"[Discovery] Phase 2 Serverless enrichment complete")
         except Exception as e:
             logger.error(f"[Discovery] Phase 2 Serverless enrichment failed: {e}")
-            summary["errors"].append(f"Serverless enrichment failed: {str(e)}")
+            error_msg = f"Serverless enrichment failed: {str(e)}"
```

Ruff (0.15.12)

[warning] Use explicit conversion flag

Replace with conversion flag

(RUF010)